File: src-tauri/src/ssh.rs
/**
 * SSH 连接相关命令处理
 */

use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::sync::{Arc, Mutex};

/// SSH 连接信息
#[derive(Debug, Clone)]
pub struct SshConnection {
    pub server_id: String,
    pub host: String,
    pub port: u16,
    pub username: String,
    // 注意：实际实现中不应该存储密码，应该使用密钥或会话管理
}

/// 全局连接池
type ConnectionPool = Arc<Mutex<HashMap<String, SshConnection>>>;

lazy_static::lazy_static! {
    static ref CONNECTIONS: ConnectionPool = Arc::new(Mutex::new(HashMap::new()));
}

/// 连接 SSH 服务器参数
#[derive(Debug, Deserialize)]
pub struct ConnectSshParams {
    pub server_id: String,
    pub host: String,
    pub port: u16,
    pub username: String,
    pub password: Option<String>,
    pub key_path: Option<String>,
}

/// 连接 SSH 服务器返回
#[derive(Debug, Serialize)]
pub struct ConnectSshResult {
    pub success: bool,
    pub connection_id: String,
    pub message: Option<String>,
}

/// 断开 SSH 服务器参数
#[derive(Debug, Deserialize)]
pub struct DisconnectSshParams {
    pub server_id: String,
}

/// 断开 SSH 服务器返回
#[derive(Debug, Serialize)]
pub struct DisconnectSshResult {
    pub success: bool,
    pub message: Option<String>,
}

/// 执行 SSH 命令参数
#[derive(Debug, Deserialize)]
pub struct ExecuteSshCommandParams {
    pub server_id: String,
    pub command: String,
}

/// 执行 SSH 命令返回
#[derive(Debug, Serialize)]
pub struct ExecuteSshCommandResult {
    pub output: String,
    pub exit_code: i32,
}

/// 重连终端参数
#[derive(Debug, Deserialize)]
pub struct ReconnectTerminalParams {
    pub server_id: String,
}

/// 重连终端返回
#[derive(Debug, Serialize)]
pub struct ReconnectTerminalResult {
    pub success: bool,
    pub message: Option<String>,
}

/// 连接 SSH 服务器
/// 
/// # 命令名称
/// `connect_ssh_server`
/// 
/// # 参数
/// - `server_id`: 服务器ID
/// - `host`: 主机地址
/// - `port`: 端口
/// - `username`: 用户名
/// - `password`: 密码（可选）
/// - `key_path`: 密钥路径（可选）
/// 
/// # 返回
/// - `success`: 是否成功
/// - `connection_id`: 连接ID
/// - `message`: 消息（可选）
#[tauri::command]
pub async fn connect_ssh_server(params: ConnectSshParams) -> Result<ConnectSshResult, String> {
    // TODO: 实现实际的 SSH 连接逻辑
    // 这里需要使用 ssh2 或类似库来建立 SSH 连接
    
    let connection = SshConnection {
        server_id: params.server_id.clone(),
        host: params.host.clone(),
        port: params.port,
        username: params.username.clone(),
    };
    
    // 将连接添加到连接池
    let mut connections = CONNECTIONS.lock().unwrap();
    connections.insert(params.server_id.clone(), connection);
    
    Ok(ConnectSshResult {
        success: true,
        connection_id: params.server_id.clone(),
        message: Some("连接成功".to_string()),
    })
}

/// 断开 SSH 服务器连接
/// 
/// # 命令名称
/// `disconnect_ssh_server`
/// 
/// # 参数
/// - `server_id`: 服务器ID
/// 
/// # 返回
/// - `success`: 是否成功
/// - `message`: 消息（可选）
#[tauri::command]
pub async fn disconnect_ssh_server(params: DisconnectSshParams) -> Result<DisconnectSshResult, String> {
    // TODO: 实现实际的 SSH 断开逻辑
    
    let mut connections = CONNECTIONS.lock().unwrap();
    connections.remove(&params.server_id);
    
    Ok(DisconnectSshResult {
        success: true,
        message: Some("断开连接成功".to_string()),
    })
}

/// 执行 SSH 命令
/// 
/// # 命令名称
/// `execute_ssh_command`
/// 
/// # 参数
/// - `server_id`: 服务器ID
/// - `command`: 要执行的命令
/// 
/// # 返回
/// - `output`: 命令输出
/// - `exit_code`: 退出码
#[tauri::command]
pub async fn execute_ssh_command(params: ExecuteSshCommandParams) -> Result<ExecuteSshCommandResult, String> {
    // TODO: 实现实际的 SSH 命令执行逻辑
    // 这里需要使用 ssh2 或类似库来执行命令
    
    // 检查连接是否存在
    let connections = CONNECTIONS.lock().unwrap();
    if !connections.contains_key(&params.server_id) {
        return Err("服务器未连接".to_string());
    }
    
    // 模拟命令执行
    // 实际实现中应该通过 SSH 连接执行命令
    Ok(ExecuteSshCommandResult {
        output: format!("执行命令: {}\n[这是模拟输出，实际将调用 SSH 执行命令]", params.command),
        exit_code: 0,
    })
}

/// 重连终端
/// 
/// # 命令名称
/// `reconnect_terminal`
/// 
/// # 参数
/// - `server_id`: 服务器ID
/// 
/// # 返回
/// - `success`: 是否成功
/// - `message`: 消息（可选）
#[tauri::command]
pub async fn reconnect_terminal(params: ReconnectTerminalParams) -> Result<ReconnectTerminalResult, String> {
    // TODO: 实现实际的重连逻辑
    
    Ok(ReconnectTerminalResult {
        success: true,
        message: Some("重连成功".to_string()),
    })
}

File: src/api/ssh.js
/**
 * SSH 连接相关 API
 */

import { invoke } from '@tauri-apps/api/tauri'

/**
 * 连接 SSH 服务器
 * @param {Object} params - 连接参数
 * @param {string} params.serverId - 服务器ID
 * @param {string} params.host - 主机地址
 * @param {number} params.port - 端口
 * @param {string} params.username - 用户名
 * @param {string} [params.password] - 密码（可选）
 * @param {string} [params.keyPath] - 密钥路径（可选）
 * @returns {Promise<{success: boolean, connectionId: string}>}
 */
export async function connectSshServer(params) {
  try {
    const result = await invoke('connect_ssh_server', {
      params: {
        server_id: params.serverId,
        host: params.host,
        port: params.port,
        username: params.username,
        password: params.password || null,
        key_path: params.keyPath || null
      }
    })
    return result
  } catch (error) {
    console.error('连接SSH服务器失败:', error)
    throw new Error(error.message || '连接服务器失败')
  }
}

/**
 * 断开 SSH 服务器连接
 * @param {string} serverId - 服务器ID
 * @returns {Promise<{success: boolean}>}
 */
export async function disconnectSshServer(serverId) {
  try {
    const result = await invoke('disconnect_ssh_server', {
      params: {
        server_id: serverId
      }
    })
    return result
  } catch (error) {
    console.error('断开SSH服务器连接失败:', error)
    throw new Error(error.message || '断开连接失败')
  }
}

/**
 * 执行 SSH 命令
 * @param {Object} params - 命令参数
 * @param {string} params.serverId - 服务器ID
 * @param {string} params.command - 要执行的命令
 * @returns {Promise<{output: string, exitCode: number}>}
 */
export async function executeSshCommand(params) {
  try {
    const result = await invoke('execute_ssh_command', {
      params: {
        server_id: params.serverId,
        command: params.command
      }
    })
    return result
  } catch (error) {
    console.error('执行SSH命令失败:', error)
    throw new Error(error.message || '执行命令失败')
  }
}

/**
 * 重连终端
 * @param {string} serverId - 服务器ID
 * @returns {Promise<{success: boolean}>}
 */
export async function reconnectTerminal(serverId) {
  try {
    const result = await invoke('reconnect_terminal', {
      params: {
        server_id: serverId
      }
    })
    return result
  } catch (error) {
    console.error('重连终端失败:', error)
    throw new Error(error.message || '重连失败')
  }
}

